analyzers/src/SonarAnalyzer.Common/Helpers/SonarLintXmlReader.cs

```diff
@@ -0,0 +1,106 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2023 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.IO;
+using System.Xml.Serialization;
+using Microsoft.CodeAnalysis.Text;
+
+namespace SonarAnalyzer.Helpers;
+
+public class SonarLintXmlReader
+{
+    private readonly SonarLintXml sonarLintXml;
+    private readonly Lazy<SonarLintXmlSettingsReader> settings;
+
+    public static readonly SonarLintXmlReader Empty = new(null, LanguageNames.CSharp);
+
+    public SonarLintXmlSettingsReader Settings => settings.Value;
+
+    public SonarLintXmlReader(SourceText originalXml, string language)
+    {
+        sonarLintXml = originalXml == null ? SonarLintXml.Empty : ParseContent(originalXml);
+        settings = new Lazy<SonarLintXmlSettingsReader>(() => new(sonarLintXml, language));
+    }
+
+    private static SonarLintXml ParseContent(SourceText sonarProjectConfig)
+    {
+        try
+        {
+            var serializer = new XmlSerializer(typeof(SonarLintXml));
+            using var sr = new StringReader(sonarProjectConfig.ToString());
+            return (SonarLintXml)serializer.Deserialize(sr);
```

Comment:
This might convert from UTF-8 (most XML files are UTF-8 encoded) to UTF-16 under the hood. You can avoid this by using a combination of
* https://learn.microsoft.com/en-us/dotnet/api/microsoft.codeanalysis.text.sourcetext.write?view=roslyn-dotnet
* https://learn.microsoft.com/en-us/dotnet/api/system.io.streamwriter.-ctor?view=net-7.0#system-io-streamwriter-ctor(system-io-stream)
* https://learn.microsoft.com/en-us/dotnet/api/microsoft.codeanalysis.text.sourcetext.encoding?view=roslyn-dotnet
* https://learn.microsoft.com/en-us/dotnet/api/system.io.streamreader.-ctor?view=net-7.0#system-io-streamreader-ctor(system-io-stream-system-text-encoding-system-boolean)
* And a MemoryStream as an underlying stream